research/design-doc/sparse_checkout_asishkumar.md

```diff
@@ -0,0 +1,40 @@
+# KPM sparse checkout
+
+**Author**: Asish Kumar
+
+## Abstract
+
+`kpm` manages third-party libraries through Git repositories, requiring a `kcl.mod` file at the root directory. It treats the entire Git repository as a single `kcl` package, which is inefficient for monorepos containing multiple `kcl` packages. Often, a `kcl` project depends on just one package within a monorepo, but `kpm` downloads the entire repository. Therefore, `kpm` needs to allow adding a subdirectory of a Git repository as a dependency, enabling it to download only the necessary parts and improve performance.
+
+## User Interface
+
+The user can just provide the subdir git url. An example command will look like this:
+
+```
+kcl mod add --git https://github.com/kcl-lang/modules/tree/main/argoproj --tag <tag>
+```
+
+kpm would parse the git url and extract the subdirectory path using `GetPath()` function from github.com/kubescape/go-git-url package. It will then download the subdirectory and append it in the subdir array of the `kcl.mod` file. 
+
+The `kcl.mod` file will look like this:
+
+```
+[dependencies]
+bbb = { git = "https://github.com/kcl-lang/modules", commit = "ade147b", subdir = ["add-ndots"]}
+```
+
+The subdir is a list because in the future if user wants to add another subdir from the same git repo then it can be added without overwritting the current subdir.
```

Comment:
There are still some things I don't understand here.

> then it can be added without overwritting the current subdir

If we refer to different directories in the same repository, I think the `kcl.mod` will looks like:

```
[dependencies]
bbb = { git = "https://github.com/kcl-lang/modules/add-ndots", commit = "ade147b"}
another_bbb = { git = "https://github.com/kcl-lang/modules/another-add-ndots", commit = "ade147b"}
```

And, the two packages are stored in two different directories.

So can you show an example to describe the problem you mentioned above?

> The subdir is a list because in the future if user wants to add another subdir from the same git repo then it can be added without overwritting the current subdir.

Reply:
In my proposal, I am trying to achieve this directory structure : 

├── .kcl
│   ├── kpm
│   │   ├── modules
│   │   │   ├── k8s
│   │   │   ├── agent

> And, the two packages are stored in two different directories.
Do you mean this directory structure?

├── .kcl
│   ├── kpm
│   │   ├──modules
│   │   │   └── k8s
│   │   ├── modules
│   │   │   ├── agent

Reply:
The two file trees you gave above, are they different on the file system 😳 ? They seem to be just different drawings of the same file tree.

> In my proposal, I am trying to achieve this directory structure :

The directory structure is correct and I have no problem. My question is, why is there a `subdir` field in `kcl.mod`, and why is it a `list`. Is there a problem with some mechanism of `go-getter` ?

Reply:
> The two file trees you gave above, are they different on the file system?

Sorry, I misunderstood something. 

> why is there a subdir field in kcl.mod, and why is it a list. Is there a problem with some mechanism of go-getter ?

No, there is no problem with go-getter. I have removed subdir flag. Thanks for the suggestion!

Reply:
https://github.com/kcl-lang/kpm/pull/335#issuecomment-2151338180

Should I revert back my subdir/package change in kcl.mod?

Reply:
Hi @officialasishkumar 

I think you can follow the `cargo` design for this feature. I have one question about your design: why do you want to use a list as the `subdir`? You should also add some details about how users can import dependencies in KCL code.

Reply:
I have removed subdir. I think i will be able to parse and download the dependencies by just keeping the whole url.

Reply:
Hi @officialasishkumar 😄, I think there may be some misunderstandings between us. What needs to be determined in this document is how users can use this function properly. I don't worry about whether this function can be implemented. This feature does not involve some complex and hard parts, and there should be no obstacles to implementation. Maybe you could consider the option I mentioned earlier https://github.com/kcl-lang/kpm/pull/335#issuecomment-2151338180